slides/08_shiny.qmd

```diff
@@ -0,0 +1,442 @@
+- Don't interrupt reactivity
+- Preferably use
+  - `reactive` over `reactiveValues` 
+  - `observeEvent` over `observe`
+
+## Reactivity: Resolve early & validate generously
+
+```{r, echo = TRUE}
+#| code-line-numbers: "3-6,8-13"
+server <- function(input, output, session) {
+  output$plot <- renderPlot({
+    # resolve reactive values
+    xvar <- input$xvar
+    yvar <- input$yvar
+    req(xvar, yvar)
+    
+    # validate – is it likely that the plot will be meaningful?
+    validate(
+      need(xvar %in% names(df), glue("xvar \"{xvar}\" does not exist in data")),
+      need(yvar %in% names(df), glue("yvar \"{yvar}\" does not exist in data")),
+      need(nrow(df) > 3, "too few data points for meaningful plot")
+    )
+    
+    # write "business logic" in a unit testable function
+    my_special_plot(df[[xvar]], df[[yvar]])
+  })
+}
+```
+
+## Reactivity: Debugging with global variables
+
+```{r, echo = TRUE}
+#| code-line-numbers: "16"
+server <- function(input, output, session) {
+  output$plot <- renderPlot({
+    # resolve reactive values
+    xvar <- input$xvar
+    yvar <- input$yvar
+    req(xvar, yvar)
+    
+    # validate – is it likely that the plot will be meaningful?
+    validate(
+      need(xvar %in% names(df), glue("xvar \"{xvar}\" does not exist in data")),
+      need(yvar %in% names(df), glue("yvar \"{yvar}\" does not exist in data")),
+      need(nrow(df) > 3, "too few data points for meaningful plot")
+    )
+    
+    # write "business logic" in a unit testable function
+    .GlobalEnv$xvar <- xvar; .GlobalEnv$yvar <- yvar;  stop()
+    my_special_plot(df[[xvar]], df[[yvar]])
+  })
+}
+```
+
+## Reactivity: "Stringy" reactivity graph
+
+::: columns
+::: {.column width="50%"}
+Bad
+![](resources/connected.png){fig-align="center"}
+Many edges, difficult to anticipate reactivity behaviour
+:::
+ 
+::: {.column width="50%"}
+Good
+![](resources/stringy.png){fig-align="center"}
+This "stringy" reactivity graph is easier to understand and debug
+:::
+:::
+
+# Testing
+
+## Testing: From Icecream to Pyramid
+
+::: columns
+::: {.column width="50%"}
+Bad
+
+![](resources/icecream.png){height=400}
+
+- cannot always test manually
+- can easily introduce bugs
+:::
+ 
+::: {.column width="50%"}
+Good
+
+![](resources/pyramid.png){height=400}
+
+- manual testing is reduced
+- only for "playing around"
+:::
+:::
+
+## Testing: Snapshot tests for UI
+
+::: columns
+::: {.column width="50%"}
+```{r, echo = TRUE, eval = FALSE}
+#| code-line-numbers: "5-9"
+test_that("myInput UI works", {
+  input <- "foo"
+  set.seed(123)
+  datasets <- mock_datasets()
+  expect_snapshot(myInput(
+    "my_test",
+    datasets = datasets,
+    input = input
+  ))
+})
+```
+:::
+
+::: {.column width="50%"}
+- Shiny UI functions return shiny tag objects, i.e. HTML code
+- Can use snapshot tests to avoid accidental changes
+- Value is limited
+- Can create issues sometimes (non-reproducible hashes)
+:::
+:::
+
+## Testing: Shiny server tests
```

Comment:
maybe worth mentioning the `shinytest` package here

Reply:
coming later

Reply:
but this "testing" part is ahead of the shinytest right? I mean, when introducing the tests, shall we introduce the tools first?